### PR TITLE
Migrating android code to Kotlin

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Not entirely migrated yet, but I'm working on it. It compiles and run, but I
have to verify that everything migrated makes sense.

This is separated in two commits, one for move the files, and other for migrate to kotlin, so reading the second commit, is easier to follow.

### CODE REVIEW

**Comment by jespino** [comment]
@enahum this is a work in progress experiment, because I have been learning android and kotlin lately, so... This is perfectly doable gradually, so we can add the build.gradle changes to allow to use kotlin and migrate file by file. or even, if you don't see much value, we can avoid the migration entirely.

**Comment by enahum** [comment]
cool Jesús thanks for this. Not priority though. so it may never get in :P

**Comment by jespino** [comment]
@enahum yes, I know, this only makes sense if provides any value and don't generates much disruption. This is in part a learning exercise for me, so, no expectations here about get this merged or not.

**Comment by enahum** [comment]
@jespino I know nothing about Kotlin really, can you help enumerate what are the pros and cons of doing this? maybe we can consider it for v2. Maybe add it to the shared doc for Mobile v2 requirements?

**Comment by jespino** [comment]
@enahum the main advantages is about having a more modern language to work on. Have some degree of type inference, way less boilerplate code, supporting things like singleton classes, data classes, functions as first class citizens, implementing Single Method Interface with simple clousures, couroutines support (not fully official yet, but probably soon), not null value (unless you explicitly say that), not need to define getters and setters (and automatically access getter/setters as properties on java libraries).

Other point to take it into consideration is now is Android is betting really hard on Kotlin and is pretty well supported.

Kotlin for android includes some cool things if you are devleoping native apps, like in Activities you have all the Views objects linked by ids as properties of the activity (magically).

The list of advantages is pretty long, but at the end of the day is just a modern language, you can see this PR to see things that makes it more terse.

File: android/app/src/main/java/com/mattermost/react_native_interface/KeysReadableArray.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package com.mattermost.react_native_interface
```
Comment (jespino) [inline android/app/src/main/java/com/mattermost/react_native_interface/KeysReadableArray.kt:1]: This file was 68 lines before, now is just 30.

File: android/app/src/main/java/com/mattermost/react_native_interface/ResolvePromise.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package com.mattermost.react_native_interface
```
Comment (jespino) [inline android/app/src/main/java/com/mattermost/react_native_interface/ResolvePromise.kt:1]: before 64 lines, now 21

**Comment by jespino** [comment]
I going to close this proof of concept because is no longer interesting. The mobile V2 should decide what to use for the language.

**Pull request closed without merging**